src/compiler/checker.ts

```diff
@@ -32527,8 +32560,10 @@ namespace ts {
                 let wouldWorkWithAwait = false;
                 const errNode = errorNode || operatorToken;
                 if (isRelated) {
-                    const awaitedLeftType = getAwaitedType(leftType);
-                    const awaitedRightType = getAwaitedType(rightType);
+                    let awaitedLeftType = getAwaitedType(leftType);
+                    let awaitedRightType = getAwaitedType(rightType);
+                    awaitedLeftType &&= unwrapAwaitedType(awaitedLeftType);
+                    awaitedRightType &&= unwrapAwaitedType(awaitedRightType);

```

Comment:
I don't think I understand why we'd unwrap here, and I doubt this has test coverage since it would only have an effect in higher order, and the test cases I recall are very simple `3 + promiseOfNumber` kind of operations. Actually, off the top of my head, I’m not sure you could come up with a test case where this matters, since it’s for binary operators where `wouldWorkWithAwait` would only come out true if the awaited types were primitives?

Reply:
I had to add this or we were getting `Did you forget to use 'await'?` related errors for every invalid expression like `1 + {}`. That may have been remediated by restrictions I added around when we add `Awaited<T>` to the result of `getAwaitedType`, but in general I think its a good idea to leave this in since this logic previously expected `getAwaitedType` to return `T` (and not `Awaited<T>`).

Reply:
The problem is with the behavior of assignability with respect to recursive conditional types. Take these two examples:

**Example 1**
```ts
// @strict: true

var v: void;
var b: boolean;
function f<T>(x: T, y: Awaited<T>) {
    x < v; // error
    y < v; // ok

    x < b; // error
    y < b; // error
}
```
[Playground Link](https://www.staging-typescript.org/play?ts=4.5.0-pr-45350-27&ssl=1&ssc=17&pln=1&pc=13#code/PTAEAEGcBcCcEsDG0Bco4FcCmAoHA3AQ1lHzXwHt4ATAbgONACM0mKKAbLQgO3oDMMPZPAo9Q-ADwAVAHwAKAB5ppAGlABPNAEEA7oXjQs1GbICUoAN45Qt0ItCTS9O5sfO8rh06Yu7G918cAF8gA)

**Example 2**:
```ts
// @strict: false

var v: void;
var b: boolean;
function f<T>(x: T, y: Awaited<T>) {
    x < v; // error
    y < v; // ok

    x < b; // error
    y < b; // ok
}
```
[Playground Link](https://www.staging-typescript.org/play?strict=false&noImplicitAny=false&strictNullChecks=false&strictFunctionTypes=false&strictPropertyInitialization=false&strictBindCallApply=false&noImplicitThis=false&noImplicitReturns=false&alwaysStrict=false&ts=4.5.0-pr-45350-27#code/PTAEAEGcBcCcEsDG0BcoBmBDANpApgFAEBumsoxaxA9vACYDcJZoARmq9ddnpgHZN0AVz7J41PhgA8AFQB8ACgAeaGQBpQATzQBBAO6Z40PHVlyAlKADeBUHdBLQUik3tanLom8fPWr+5oefgQAvkA)

In Example 1 (strict mode), `x < v` is an error, but `y < v` is *not* an error, which results in `wouldWorkWithAwait` being `true`. It also behaves that way if `v` is `any[]`, for some reason.

In Example 2 (non-strict mode), `x < v` and `x < b` are errors, but `y < v` and `y < b` are *not* errors, which results in  `wouldWorkWithAwait` being `true` for *every* failed comparison operation.

As a result, I think this needs to stay, but I'll do a little more digging.

Reply:
When we're checking whether `Awaited<T>` is comparable to `void`, we get to a point where we're checking whether the default constraint of `Awaited<T>` is comparable to `void`. The default constraint of `Awaited<T>` is `T | (T extends object ? T extends { then(onfulfilled: infer F): any; } ? F extends (value: infer V) => any ? Awaited<V> : never : T : T)`, so we end up in a union comparison.

In the union comparison we use `someTypeRelatedToType` because we're using the comparable relation. During this, we check if `T` is related to `void`, and in the process call `getNormalizedType(T, /*forWriting*/ false)` which returns the `null` type, which is comparable to `void`. This means `Awaited<T>` is comparable to `void`, so `wouldWorkWithAwait` ends up `true`.

Reply:
Also, you don't trigger the suggestion diagnostic because `getAwaitedType` for `Awaited<T>` produces `Awaited<T>`, so the type doesn't change (and thus won't trigger the related error diagnostic).